src/synth/fluid_synth_polymono.c

```diff
@@ -0,0 +1,536 @@
+						break;
+					}
+				}
+			}	
+		}
+
+		/* last_end_range: next basic channel  or midi_channels count  */
+		for (last_end_range = basicchan +1; last_end_range < n_chan; last_end_range++)
+		{
+			if (synth->channel[last_end_range]->mode &  FLUID_CHANNEL_BASIC) break;
+		}
+		/* Now last_begin_range is set */
+		switch (mode = mode &  FLUID_CHANNEL_MODE_MASK)
+		{
+			case FLUID_CHANNEL_MODE_OMNION_POLY:	/* Mode 0 and 1 */
+			case FLUID_CHANNEL_MODE_OMNION_MONO:
+				last_begin_range = last_end_range;
+				break;
+			case FLUID_CHANNEL_MODE_OMNIOFF_POLY:		/* Mode 2 */
+				last_begin_range = basicchan + 1;
+				break;
+			case FLUID_CHANNEL_MODE_OMNIOFF_MONO:		/* Mode 3 */
+				if (val) last_begin_range = basicchan + val;
+				else last_begin_range = last_end_range;
+		}
+		/* Check if val overlaps the next basic channel */
+		if (last_begin_range > last_end_range)
+		{	
+			/* val have number of channels that overlaps the next basic channel */
+			FLUID_LOG(FLUID_INFO,warning_msg2,basicchan);
+			return FLUID_FAILED;
+		}
+		/* if previous basic channel exists, val is narrowed */
+		if(prev_basic_chan >= 0)
+		{
+			synth->channel[prev_basic_chan]->mode_val = basicchan - prev_basic_chan;
+			FLUID_LOG(FLUID_INFO,warning_msg1,prev_basic_chan,basicchan - prev_basic_chan);
+		}
+
+		/* val is limited up to LastBeginRange */
+		val = last_begin_range - basicchan;
+		/* Set the Mode to the range zone: Beginning range + Ending range */
+		for (i = basicchan; i < last_end_range; i++)
+		{	
+			int new_mode = mode; /* OMNI_OFF/ON, MONO/POLY ,others bits are zero */
+			/* MIDI specs: when mode is changed, channel must receive
+			 ALL_NOTES_OFF */
+			fluid_synth_all_notes_off_LOCAL (synth, i);
+			/* basicchan only is marked Basic Channel */
+			if (i == basicchan)	new_mode |= FLUID_CHANNEL_BASIC; 
+			else val =0; /* val is 0 for other channel than basic channel */
+			/* Channel in beginning zone are enabled */
+			if (i < last_begin_range) new_mode |= FLUID_CHANNEL_ENABLED; 
+			/* Channel in ending zone are disabled */
+			else new_mode = 0;
+			/* Now mode is OMNI OFF/ON,MONO/POLY, BASIC_CHANNEL or not
+			   ENABLED or not */	
+			set_fluid_channel_basic_channel_infos(synth->channel[i],new_mode);
+			synth->channel[i]->mode_val = val;
+		}
+	}
+	return result;
+}
+
+/**
+ * Returns poly mono mode informations from any MIDI channel.
+ *
+ * @param synth the synth instance
+ * @param chan MIDI channel number (0 to MIDI channel count - 1)
+ * @param modeInfos Pointer to a #fluid_basic_channel_infos_t struct
+ *
+ * @return
+ * - FLUID_OK on success.
+ * - FLUID_FAILED 
+ *   - synth is NULL.
+ *   - chan is outside MIDI channel count.
+ *   - modeInfos is NULL.
+ */
+int fluid_synth_get_channel_mode(fluid_synth_t* synth, int chan,
+				fluid_basic_channel_infos_t  *modeInfos)
+{
+	/* check parameters first */
+	fluid_return_val_if_fail (modeInfos!= NULL, FLUID_FAILED);
+	FLUID_API_ENTRY_CHAN(FLUID_FAILED);
+	/**/
+	modeInfos->basicchan= chan;
```

Comment:
Why are you setting `chan` as basic channel? Wouldnt you need to search to what basic channel `chan` belongs to and return this instead?

Reply:
>Wouldnt you need to search to what basic channel chan belongs to and return this instead?.

No, we need only informations on this particular channel (regardless of the basic channel it belongs). I admit that the used return structure name `fluid_basic_channel_infos_t` is improper and confusing. 
Should we define a new named structure for this API ?  or simply could we better document the API ?

Reply:
No not introducing yet another struct.

I still dont understand why you dont return the basic channel `chan` belongs to. There is no gain in information returning `chan` for this member. In fact I find it more confusing. The caller knows anyway with what `chan` the function was called.

Reply:
>I still dont understand why you dont return the basic channel chan belongs to.

Ok, it will be done.